url748e34f967_Diab1o_java_7z_archiver_tgz-pJ8-com_swemel_sevenzip_TestJ8/src/com/swemel/sevenzip/compression/lzma/Encoder.java

```diff
@@ -0,0 +1,1316 @@
+                repMaxIndex = i;
+            }
+        }
+        if (repLens[repMaxIndex] >= _numFastBytes) {
+            backRes = repMaxIndex;
+            int lenRes = repLens[repMaxIndex];
+            movePos(lenRes - 1);
+            return lenRes;
+        }
+
+        if (lenMain >= _numFastBytes) {
+            backRes = _matchDistances[numDistancePairs - 1] + Base.kNumRepDistances;
+            movePos(lenMain - 1);
+            return lenMain;
+        }
+
+        byte currentByte = _matchFinder.getIndexByte(0 - 1);
+        byte matchByte = _matchFinder.getIndexByte(0 - _repDistances[0] - 1 - 1);
+
+        if (lenMain < 2 && currentByte != matchByte && repLens[repMaxIndex] < 2) {
+            backRes = -1;
+            return 1;
+        }
+
+        _optimum[0].State = state;
+
+        int posState = (position & _posStateMask);
+
+        _optimum[1].Price = com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice0(_isMatch[(state << Base.kNumPosStatesBitsMax) + posState]) +
+                _literalEncoder.getSubCoder(position, _previousByte).getPrice(!Base.stateIsCharState(state), matchByte, currentByte);
+        _optimum[1].makeAsChar();
+
+        int matchPrice = com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice1(_isMatch[(state << Base.kNumPosStatesBitsMax) + posState]);
+        int repMatchPrice = matchPrice + com.swemel.sevenzip.compression.rangecoder.Encoder.getPrice1(_isRep[state]);
+
+        if (matchByte == currentByte) {
+            int shortRepPrice = repMatchPrice + getRepLen1Price(state, posState);
+            if (shortRepPrice < _optimum[1].Price) {
+                _optimum[1].Price = shortRepPrice;
+                _optimum[1].makeAsShortRep();
+            }
+        }
+
+        int lenEnd = ((lenMain >= repLens[repMaxIndex]) ? lenMain : repLens[repMaxIndex]);
+
+        if (lenEnd < 2) {
+            backRes = _optimum[1].BackPrev;
+            return 1;
+        }
+
+        _optimum[1].PosPrev = 0;
+
+        _optimum[0].Backs0 = reps[0];
+        _optimum[0].Backs1 = reps[1];
+        _optimum[0].Backs2 = reps[2];
+        _optimum[0].Backs3 = reps[3];
+
+        int len = lenEnd;
+        do {
+            _optimum[len--].Price = kIfinityPrice;
+        }
+        while (len >= 2);
+
+        for (i = 0; i < Base.kNumRepDistances; i++) {
+            int repLen = repLens[i];
+            if (repLen < 2) {
+                continue;
+            }
+            int price = repMatchPrice + getPureRepPrice(i, state, posState);
+            do {
+                int curAndLenPrice = price + _repMatchLenEncoder.getPrice(repLen - 2, posState);
+                Optimal optimum = _optimum[repLen];
+                if (curAndLenPrice < optimum.Price) {
+                    optimum.Price = curAndLenPrice;
+                    optimum.PosPrev = 0;
+                    optimum.BackPrev = i;
+                    optimum.Prev1IsChar = false;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 632:637, 848:853.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.